pkg/smbios/table.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2016-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package smbios
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// Table is a generic type of table that does not parsed fields,
+// it only allows access to its contents by offset.
+type Table struct {
+	Header
+	data    []byte   `smbios:"-"` // Structured part of the table.
+	strings []string `smbios:"-"` // Strings section.
+}
+
+var (
+	// ErrTableNotFound is retuned if table with specified type is not found.
+	ErrTableNotFound = errors.New("table not found")
+
+	// ErrUnsupportedTableType is returned by ParseTypedTable if this table type is not supported and cannot be parsed.
+	ErrUnsupportedTableType = errors.New("unsupported table type")
+
+	errEndOfTable = errors.New("end of table")
+)
+
+// Len returns length of the structured part of the table.
+func (t *Table) Len() int {
+	return len(t.data)
+}
+
+// GetByteAt returns a byte from the structured part at the specified offset.
+func (t *Table) GetByteAt(offset int) (uint8, error) {
+	if offset > len(t.data)-1 {
+		return 0, fmt.Errorf("invalid offset %d", offset)
+	}
+	return t.data[offset], nil
+}
+
+// GetBytesAt returns a number of bytes from the structured part at the specified offset.
+func (t *Table) GetBytesAt(offset, length int) ([]byte, error) {
+	if offset > len(t.data)-length {
+		return nil, fmt.Errorf("invalid offset %d", offset)
+	}
+	return t.data[offset : offset+length], nil
+}
+
+// GetWordAt returns a 16-bit word from the structured part at the specified offset.
+func (t *Table) GetWordAt(offset int) (res uint16, err error) {
+	if offset > len(t.data)-2 {
+		return 0, fmt.Errorf("invalid offset %d", offset)
+	}
+	buf := bytes.NewBuffer(t.data[offset : offset+2])
+	binary.Read(buf, binary.LittleEndian, &res)
+	return res, nil
+}
+
+// GetDWordAt returns a 32-bit word from the structured part at the specified offset.
+func (t *Table) GetDWordAt(offset int) (res uint32, err error) {
+	if offset > len(t.data)-4 {
+		return 0, fmt.Errorf("invalid offset %d", offset)
+	}
+	buf := bytes.NewBuffer(t.data[offset : offset+4])
+	binary.Read(buf, binary.LittleEndian, &res)
+	return res, nil
+}
+
+// GetQWordAt returns a 64-bit word from the structured part at the specified offset.
+func (t *Table) GetQWordAt(offset int) (res uint64, err error) {
```

Comment:
This stuff seems a lot like uio.Lexer: https://godoc.org/github.com/u-root/u-root/pkg/uio#Lexer

Reply:
well, yes and no. yes, it is reading different sized fields from a buffer, but there's also string with its weird index indirection thing. also "byte", "word", "dword" and "qword" are terms from the spec, so i think it will help to have them in the interface instead of 8/16/32/64.
so, i'm keeping the interface. but i have simplified the code a bit by using NewReader.

Reply:
Idk, you just convert reflect.Uint8,16,32,64 to these.

How about I add aliases to uio?